### PR TITLE
Fix sidecar DNS proxying with dual stack

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/51608

### CODE REVIEW

File: pilot/pkg/model/service.go [hunk 1]
```diff
@@ -1272,6 +1272,22 @@ func (s *Service) GetExtraAddressesForProxy(node *Proxy) []string {
 	return nil
 }
 
+// GetAllAddressesForProxy returns a k8s service's extra addresses to the cluster where the node resides.
+// Especially for dual stack k8s service to get other IP family addresses.
+func (s *Service) GetAllAddressesForProxy(node *Proxy) []string {
+	if features.EnableDualStack && node.Metadata != nil {
+		if node.Metadata.ClusterID != "" {
```
Comment (leosarra) [inline pilot/pkg/model/service.go:1279]: would it make sense to fallback to getAddressForProxy(node) if node.Metadata.ClusterID == "" ?

**Comment by leosarra** [comment]
Overall LGTM. Would be possible to add a test case for this?

**Comment by howardjohn** [comment]
> Overall LGTM. Would be possible to add a test case for this?

Yep, i was meaning to but ran out of time. Just added + marked ready

File: pkg/dns/server/name_table.go [hunk 2]
```diff
@@ -42,17 +42,22 @@ func BuildNameTable(cfg Config) *dnsProto.NameTable {
...
 			// should be caught in validation.
 			if !netutil.IsValidIPAddress(svcAddress) {
 				continue
 			}
 			addressList = append(addressList, svcAddress)
-		} else {
+		}
+		if headless {
 			// The IP will be unspecified here if its headless service or if the auto
 			// IP allocation logic for service entry was unable to allocate an IP.
```
Comment (hzxuzhonghu) [inline pkg/dns/server/name_table.go:62]: What does this mean

**Review by hzxuzhonghu — APPROVED** [review]
Now GetXXAddressXXs are a little bit too confusing

**Pull request merged**